### ISSUE TITLE
Wrong crossentropy loss calculation

### ISSUE THREAD

**Issue by mayurnewase** [pytorch/pytorch#25571]
## 🐛 Bug

I tried nn.CrossentropyLoss() using weights with reduction of none, sum, mean.None and sum is working as expected but with mean loss is wrong

## To Reproduce
`
suppose 3 classes

pred = torch.tensor([[0.2, 0.3, 0.5], [ 0.5, 0.2, 0.3 ]])
y =  torch.tensor([2, 1])
weights = torch.tensor([2., 4., 6.])

loss_function = nn.CrossEntropyLoss(weight = weights, size_average = None, ignore_index = -100, reduce = None, reduction="**none**")
#gives (tensor([5.6390, 4.9593]))

loss_function = nn.CrossEntropyLoss(weight = weights, size_average = None, ignore_index = -100, reduce = None, reduction="**sum**")
#gives (tensor(10.5983)

but 
loss_function = nn.CrossEntropyLoss(weight = weights, size_average = None, ignore_index = -100, reduce = None, reduction="**mean**")
#gives tensor(1.0598)

## Expected behavior
mean((5.6390, 4.9593)) = 5.29915

## Environment
pytorch 1.2.0 with cuda

**Comment by NekuSakuraba** [comment]
Hello!

I tried to understand and reproduce the problem.
Actually, I don't know if it's a problem because you passed weights, so it gave to us a weighted average.

The first observation had a weight of 6 and the second one had 4,
so (5.6390 + 4.9593)/(6+4) = 1.0598.

**Comment by soumith** [comment]
closing via Neku's comment

**Comment by mayurnewase** [comment]
I did not understand your calculation.
Formula for crossentropy with weights from docs is 
`loss(x,class)=weight[class][−log( exp(x[class]) / ∑exp(x[j]) )]`
so in this case i think it should be
`a = 6 * (-np.log(np.exp(0.5) / ∑exp(x[j])))
b = 4 * (-np.log(np.exp(0.2) / ∑exp(x[j])))`
**reduction: sum**
`a+b`
**reduction:mean**
`a+b/2`
sum is correct here,mean is not

**Comment by NekuSakuraba** [comment]
Since you passed the weights, the mean should be a weighted mean.

Let's say...
`a = (-np.log(np.exp(0.5) / ∑exp(x[j])))`
`b = (-np.log(np.exp(0.2) / ∑exp(x[j])))`

Then...
**sum**
`6*a + 4*b`
**mean**
`(6*a + 4*b)/(6+4)`

**Comment by mayurnewase** [comment]
Oh gotcha...
Thanks
